api/src/task/notifications/generate_notifications.py

```diff
@@ -92,357 +53,55 @@ def __init__(
-                    len(container.saved_opportunities)
-                    for container in self.user_notification_map.values()
-                ),
-            },
-        )
-
-    def _collect_search_notifications(self) -> None:
-        """Collect notifications for changed saved searches"""
-        # Get all saved searches that haven't been checked since last notification
-        stmt = select(UserSavedSearch).where(
-            UserSavedSearch.last_notified_at < datetime_util.utcnow()
-        )
-        saved_searches = self.db_session.execute(stmt).scalars()
-
-        # Group searches by query to minimize search index calls
-        query_map: dict[str, list[UserSavedSearch]] = {}
-        for saved_search in saved_searches:
-            # Remove pagination parameters before using as key
-            search_query = _strip_pagination_params(saved_search.search_query)
-            query_key = str(search_query)
-
-            if query_key not in query_map:
-                query_map[query_key] = []
-            query_map[query_key].append(saved_search)
 
-        # For each unique query, check if results have changed
-        for searches in query_map.values():
-            current_results: list[int] = search_opportunities_id(
-                self.search_client, searches[0].search_query
-            )
+        data = OpportunityNotification(self.db_session).notification_data()
+        data and self.send_notifications(data)
 
-            for saved_search in searches:
-                previous_results = set(saved_search.searched_opportunity_ids)
-                if set(current_results) != previous_results:
-                    user_id = saved_search.user_id
-                    if user_id not in self.user_notification_map:
-                        self.user_notification_map[user_id] = NotificationContainer()
-                    self.user_notification_map[user_id].saved_searches.append(saved_search)
+        data = SearchNotification(self.db_session, self.search_client).notification_data()
+        data and self.send_notifications(data)
 
-                    # Update the saved search with new results
-                    saved_search.searched_opportunity_ids = list(current_results)
+        closing_notification = ClosingDateNotification(self.db_session)
+        data = closing_notification.notification_data()
+        data and self.send_notifications(data)
+        data and closing_notification.create_user_opportunity_notification_log()
 
-        logger.info(
-            "Collected search notifications",
-            extra={
-                "user_count": len(self.user_notification_map),
-                "total_searches": sum(
-                    len(container.saved_searches)
-                    for container in self.user_notification_map.values()
-                ),
-            },
-        )
-
-    def _send_notifications(self) -> None:
+    def send_notifications(self, data: EmailData) -> None:
```

Comment:
Why not make `send_notifications` a function in your base notification class and just have it be called after the other functions? Keep it all contained in one class, although the actual sending is a shared util